From 03a73236be9ca8aaa283675e91e0e2ac4de334d5 Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Fri, 2 Aug 2019 17:10:39 +0800
Subject: [PATCH] DRVLIB_SW-2711:Insertion-traffic-having-extra-4-bytes-fcs

    - Insertion traffic having dummy 4 bytes 0 issue is resolved.
      by setting Pmac Don't add Dummy 4 bytes and FDMA dont remove FCS
    - Dummy Special tag Insertion from MAC -> GSWIP setting is moved
      inside MAC default config, instead of gswip default config
---
 drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c | 10 +---------
 drivers/net/ethernet/lantiq/switch-api/mac/mac_cfg.c | 14 ++++++++++++++
 2 files changed, 15 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
index 8725428602a3..686fe45f30b5 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_defconf.c
@@ -213,7 +213,6 @@ int gsw_misc_config(struct core_ops *ops)
 	GSW_register_t reg;
 	int i = 0;
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(ops);
-	struct mac_ops *mac_ops;
 
 	if (gswdev == NULL) {
 		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
@@ -234,13 +233,6 @@ int gsw_misc_config(struct core_ops *ops)
 		ops->gsw_common_ops.RegisterSet(ops, &reg);
 	}
 
-	for (i = MAC_2; i < (gswdev->pnum + MAC_2); i++) {
-		mac_ops = gsw_get_mac_ops(0, i);
-
-		if (mac_ops)
-			mac_ops->mac_op_cfg(mac_ops, RX_SPTAG_INSERT);
-	}
-
 	return 0;
 }
 
@@ -533,7 +525,7 @@ static int pmac_glbl_cfg(struct core_ops *ops, u8 pmacid)
 	glbl_cfg.bJumboEna = 1;
 	glbl_cfg.nMaxJumboLen = 10000;
 	glbl_cfg.bTxFCSDis = 0;
-	glbl_cfg.bRxFCSDis = 1;
+	glbl_cfg.bRxFCSDis = 0;
 	glbl_cfg.eShortFrmChkType = GSW_PMAC_SHORT_LEN_ENA_UNTAG;
 	glbl_cfg.bLongFrmChkDis = 1;
 	glbl_cfg.bProcFlagsEgCfgEna = 1;
diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/mac_cfg.c b/drivers/net/ethernet/lantiq/switch-api/mac/mac_cfg.c
index 4e35f2ebd7dd..056c59c5fbb1 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/mac_cfg.c
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/mac_cfg.c
@@ -1085,6 +1085,7 @@ int mac_get_int_sts(void *pdev)
 int mac_init(void *pdev)
 {
 	struct mac_prv_data *pdata = GET_MAC_PDATA(pdev);
+
 #if defined(PC_UTILITY) || defined(CHIPTEST)
 	int i = 0;
 
@@ -1173,6 +1174,19 @@ int mac_init(void *pdev)
 	/* Tell adaption layer to remove FCS in Rx Direction */
 	gswss_set_mac_rxfcs_op(pdev, MODE3);
 
+	/* Insert Dummy Special tag from GSWIP Subsys
+	 * Bug Fix for Insertion was not getting hit in PCE rule
+	 * Default Insert a dummy Special Tag to all packets entering from MAC
+         * MAC -> GSWIP
+	 */
+	gswss_set_mac_rxsptag_op(pdev, (RX_SPTAG_INSERT % 4));
+
+	/* By default setting FDMA Don't Remove FCS @ Tx
+	 * PMAC is now set as no dummy FCS addition @ Tx,
+	 * Otherwise Insertion case will have 4 Extra Byte 0's
+	 */
+	gswss_set_mac_txfcs_rm_op(pdev, (TX_FCS_NO_REMOVE % 4));
+
 	/* Set XGMAC Port to MDIO Clause 22 */
 	mdio_set_clause(pdev, 1, (pdata->mac_idx - MAC_2));
 
